src/hotspot/share/gc/g1/g1RemSet.cpp

```diff
@@ -1289,21 +1289,36 @@ class G1MergeHeapRootsTask : public WorkerTask {
       Ticks start = Ticks::now();
 
       _dirty_card_buffers = NEW_C_HEAP_ARRAY(BufferNode::Stack, num_workers, mtGC);
-      for (uint i = 0; i < num_workers; i++)
-      {
+      for (uint i = 0; i < num_workers; i++) {
         new (&_dirty_card_buffers[i]) BufferNode::Stack();
       }
 
       G1DirtyCardQueueSet& dcqs = G1BarrierSet::dirty_card_queue_set();
       BufferNodeList buffers = dcqs.take_all_completed_buffers();
-      BufferNode* cur = buffers._head;
+
+      size_t num_buffers = (buffers._entry_count / G1UpdateBufferSize);
+      size_t buffers_per_thread = MAX2(num_buffers / num_workers, (size_t)1);
+
+      BufferNode* head = buffers._head;
+      BufferNode* tail = head;
```

Comment:
Because there may be non-full buffers, num_buffers may be an underestimate of
the actual number of buffers. And buffers_per_thread is rounded down. Because
of that, this calculation may result in various imbalances, which might be
somewhat significant. I think it would be better to instead calculate
(rounding up) and use entries_per_thread, and in the loop increment count by
cur->size() (before updating cur).

Reply:
Fixed